java-frontend/src/main/java/org/sonar/java/regex/JavaCharacterParser.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.regex;
+
+import org.sonar.java.regex.ast.IndexRange;
+import org.sonar.java.regex.ast.JavaCharacter;
+import org.sonar.java.regex.ast.RegexSource;
+
+/**
+ * Parse the contents of string literals and provide the individual characters of the string after processing escape
+ * sequences
+ */
+public class JavaCharacterParser {
+
+  public static final int EOF = -1;
+
+  private final RegexSource source;
+
+  private final JavaUnicodeEscapeParser unicodeProcessedCharacters;
+
+  private JavaCharacter current;
+
+  public JavaCharacterParser(RegexSource source) {
+    this.source = source;
+    this.unicodeProcessedCharacters = new JavaUnicodeEscapeParser(source);
+    moveNext();
+  }
+
+  public void moveNext() {
+    current = parseJavaCharacter();
+  }
+
+  public JavaCharacter getCurrent() {
+    return current;
+  }
+
+  public int getCurrentChar() {
+    if (current != null) {
+      return current.getCharacter();
+    } else {
+      return EOF;
+    }
+  }
+
+  public IndexRange getCurrentIndexRange() {
+    if (current != null) {
+      return current.getRange();
+    } else {
+      return new IndexRange(source.length(), source.length());
+    }
+  }
+
+  public int getCurrentStartIndex() {
+    if (current == null) {
+      return source.length();
+    } else {
+      return current.getRange().getBeginningOffset();
+    }
+  }
+
+  public boolean isAtEnd() {
+    return current == null;
+  }
+
+  public boolean isNotAtEnd() {
+    return current != null;
+  }
+
+  public boolean currentIs(char ch) {
+    return current != null && current.getCharacter() == ch;
+  }
+
+  private JavaCharacter parseJavaCharacter() {
+    JavaCharacter javaCharacter = unicodeProcessedCharacters.getCurrent();
+    if (javaCharacter == null) {
+      return null;
+    }
+    if (javaCharacter.getCharacter() == '\\') {
+      return parseJavaEscapeSequence(javaCharacter);
+    }
+    unicodeProcessedCharacters.moveNext();
+    return javaCharacter;
+  }
+
+  private JavaCharacter parseJavaEscapeSequence(JavaCharacter backslash) {
+    unicodeProcessedCharacters.moveNext();
+    JavaCharacter javaCharacter = unicodeProcessedCharacters.getCurrent();
+    if (javaCharacter == null) {
+      // Should only happen in case of syntactically invalid string literals
+      return null;
+    }
+    char ch = javaCharacter.getCharacter();
+    switch (ch) {
+      case 'n':
+        ch = '\n';
+        unicodeProcessedCharacters.moveNext();
```

Comment:
I think I would move out of the `switch` the statement `unicodeProcessedCharacters.moveNext();`, I think it simplifies a bit:

```
    switch (ch) {
      case 'n':
        ch = '\n';
        break;
      case 'r':
        ch = '\r';
        break;
      case 'f':
        ch = '\f';
        break;
      case 'b':
        ch = '\b';
        break;
      case 't':
        ch = '\t';
        break;
      default:
        if (isOctalDigit(ch)) {
          StringBuilder codeUnit = new StringBuilder(3);
          for (int i = 0; i < 3 && javaCharacter != null && isOctalDigit(javaCharacter.getCharacter()); i++) {
            codeUnit.append(javaCharacter.getCharacter());
            unicodeProcessedCharacters.moveNext();
            javaCharacter = unicodeProcessedCharacters.getCurrent();
          }
          ch = (char) Integer.parseInt(codeUnit.toString(), 8);
          return new JavaCharacter(source, backslash.getRange().extendTo(getCurrentStartIndex()), ch);
        }
    }
    unicodeProcessedCharacters.moveNext();
    return new JavaCharacter(source, backslash.getRange().extendTo(getCurrentStartIndex()), ch);
```